### PR TITLE
TS-4624: use the server UUID in the Via header

### PR DESCRIPTION
**Pull request by yatsukhnenko**
If I correctly understood the task :smiley:

### CODE REVIEW

**Comment by zwoop** [comment]
Almost, but not quite. :)  All APIs that are prefixed TS are the public APIs, and we don't use those in the internal code. Look at the log tags in logging that uses the internal APIs for the UUID. E.g. 

    const char *uuid = (char *)Machine::instance()->uuid.getString();


Makes sense?

**Comment by yatsukhnenko** [comment]
Don't quite understand why public API can't be used in internal code, but ok, changed code to `Machine::instance()->uuid.getString()`

**Comment by zwoop** [comment]
It's an inefficiency for starters (you go through a C-> C++ wrapper layer, for no good reason (remember, the public APIs are all C, whereas in the core, you can use all the C++ objects and methods directly).

In this case, it's marginal, but there are other places where it would be inefficient. And, it would add dependencies to the C-APIs in the C++ core which are undesirable in some cases.

File: proxy/http/HttpTransactHeaders.cc [hunk 1]
```diff
@@ -750,12 +752,10 @@ HttpTransactHeaders::insert_via_header_in_request(HttpTransact::State *s, HTTPHd
   via_string += nstrcpy(via_string, s->http_config_param->proxy_hostname);
 
   *via_string++ = '[';
-  /* I thought we should use the transaction local outgoing IP address but
-     that makes cycle detection (which is the point) unrealiable. We must
-     use the same value every time to be sure.
-  */
-  memcpy(via_string, Machine::instance()->ip_hex_string, Machine::instance()->ip_hex_string_len);
-  via_string += Machine::instance()->ip_hex_string_len;
+  if (uuid != NULL) {
```
Comment (zwoop) [inline proxy/http/HttpTransactHeaders.cc]: I don't think the machine uuid can ever be NULL, but harmless to check I guess :).

**Comment by zwoop** [comment]
[approve ci]

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/439/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/333/ for details.

File: proxy/http/HttpTransactHeaders.cc (hunk 1 above)
Comment (yatsukhnenko) [inline proxy/http/HttpTransactHeaders.cc]: Reinsurance :) Shoul I delete this check?

**Comment by yatsukhnenko** [comment]
Thanks for explaining about API usage

File: proxy/http/HttpTransactHeaders.cc (hunk 1 above)
Comment (zwoop) [inline proxy/http/HttpTransactHeaders.cc]: Yeah nuke it IMO. The Machine UUID is guaranteed to be initiated on startup.

Comment (yatsukhnenko) [inline proxy/http/HttpTransactHeaders.cc]: done :)

**Comment by zwoop** [comment]
[approve ci]

**Comment by atsci** [comment]
FreeBSD build *failed*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/441/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/335/ for details.

**Comment by zwoop** [comment]
I suspect this failed for other reasons, trying again [approve ci].

**Comment by zwoop** [comment]
I checked the core on this FreeBSD build, and it seems completely unrelated to pretty much anything :).

```
(gdb) bt
#0  0x0000000000000000 in ?? ()
#1  0x00000000006ec34a in EThread::process_event (this=0x804107000, e=0x803f2c880, calling_code=<value optimized out>) at I_Continuation.h:153
#2  0x00000000006ec8ec in EThread::execute (this=0x804107000) at ../../../iocore/eventsystem/UnixEThread.cc:245
#3  0x00000000006ebb82 in spawn_thread_internal (a=0x803ef5700) at ../../../iocore/eventsystem/Thread.cc:84
#4  0x00000008027514f5 in pthread_create () from /lib/libthr.so.3
#5  0x0000000000000000 in ?? ()
```

**Comment by zwoop** [comment]
@yatsukhnenko What do you think of a future feature (config) to disable this Via checking completely? If I know there's not chance of a loop, why bother checking the Via header? I guess one could just nuke the Via header on the incoming request, but there might be other reasons to keep it?

This is completely unrelated to this PR though, just thinking out loud.

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/443/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/337/ for details.

**Comment by zwoop** [comment]
Did a quick test of this, and it looks good:

    Via: http/1.1 fedora.ogre.com[f6f5162d-5a02-4d10-ac6e-35aede508fe5] (ApacheTrafficServer/7.0.0)

**Comment by yatsukhnenko** [comment]
> What do you think of a future feature (config) to disable this Via checking completely? If I know there's not chance of a loop, why bother checking the Via header?

@zwoop, makes sense. Should I create new jira for this?

**Comment by zwoop** [comment]
Yeah, file a Jira probably. I could have sworn we've had issues before where loop detection kicked in when it shouldn't (e.g. you legitimately proxy to yourself). But I can find an issue on it, maybe you can take a look too.

**Comment by yatsukhnenko** [comment]
[This](https://issues.apache.org/jira/browse/TS-2819)  one?

**Pull request merged**